blazar/enforcement/filters/external_service_filter.py

```diff
@@ -0,0 +1,124 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import json
+import requests
+
+from blazar import context
+from blazar.enforcement.filters import base_filter
+from blazar import exceptions
+from blazar.i18n import _
+from blazar.utils.openstack import base
+from blazar.utils.openstack.keystone import BlazarKeystoneClient
+
+from oslo_config import cfg
+from oslo_log import log as logging
+
+LOG = logging.getLogger(__name__)
+
+
+def date_converter(value):
+    if isinstance(value, datetime.datetime):
+        return str(value).split('.')[0]
+
+
+class ExternalServiceUnsupportedHTTPResponse(exceptions.BlazarException):
+    code = 400
+    msg_fmt = _('External Service Filter returned a %(status)s http response. '
+                'Only 204 and 403 responses are supported.')
+
+
+class ExternalServiceFilterException(exceptions.BlazarException):
+    code = 400
+    msg_fmt = _('%(message)s')
+
+
+class ExternalServiceFilter(base_filter.BaseFilter):
+
+    enforcement_opts = [
+        cfg.StrOpt(
+            'external_service_endpoint',
+            default=False,
+            help='The url of the external service API. A value of -1 will '
+                 'disabled the service.'),
+        cfg.StrOpt(
+            'external_service_token',
+            default="",
+            help='Authentication token for token based authentication.')
+    ]
+
+    def __init__(self, conf=None):
+        super(ExternalServiceFilter, self).__init__(conf=conf)
+
+    def get_headers(self):
+        if self.conf.enforcement.external_service_token:
+            return {
+                'X-Auth-Token': self.conf.enforcement.external_service_token}
+        else:
+            auth_url = "%s://%s:%s/%s" % (self.conf.os_auth_protocol,
+                                          base.get_os_auth_host(self.conf),
+                                          self.conf.os_auth_port,
+                                          self.conf.os_auth_prefix)
+            client = BlazarKeystoneClient(
+                password=self.conf.os_admin_password,
+                auth_url=auth_url,
+                ctx=context.admin())
+
+            return {'X-Auth-Token': client.auth_token}
+
+    def post(self, path, body):
+        url = self.conf.enforcement.external_service_endpoint
+
+        if url[-1] == '/':
+            url += path[1:]
+        else:
+            url += path
+
+        # Hacky way to convert dates to strings before making post
+        body = json.loads(json.dumps(body, default=date_converter))
```

Comment:
It fails because the dict object contains datetime objects. I'm surprised it just doesn't handle those in requests gracefully. The annoying thing about json dumps was that it was outputing a string, but I think this was because I was encoding twice so I'll try out using data.

One of the nice things about the json parameter is that it automatically adds `Content-Type: application/json` to the header, but that's an easy fix.

Reply:
Fixed.